django/db/migrations/optimizer.py

```diff
@@ -44,20 +44,29 @@ def optimize_inner(self, operations, app_label=None):
         """
         new_operations = []
         for i, operation in enumerate(operations):
+            right = True  # Should we reduce on the right or on the left.
             # Compare it to each operation after it
             for j, other in enumerate(operations[i + 1:]):
                 in_between = operations[i + 1:i + j + 1]
-                result = operation.reduce(other, in_between, app_label)
+                result = operation.reduce(other, app_label)
                 if isinstance(result, list):
-                    # Optimize! Add result, then remaining others, then return
-                    new_operations.extend(result)
-                    new_operations.extend(in_between)
+                    if right:
+                        new_operations.extend(in_between)
+                        new_operations.extend(result)
+                    elif all(op.reduce(other, app_label) is True for op in in_between):
+                        # Perform a left reduction if all the in between
+                        # operations can optimize through other.
+                        new_operations.extend(result)
+                        new_operations.extend(in_between)
```

Comment:
> I think the reduction operations currently assume that things will always be added to the front of the operation list

I don't think so, `optimize_inner` keeps returning as soon as an optimization is performed.

> so keeping that consistent might prevent weird optimizer loops like
[A, B] -> [B, A] -> [A, B] -> [B, A].

For that to happen you'd need to have `[A, ..., B, ...]` such that `A.reduce(B) -> [B, A]` and `B.reduce(A) -> [A, B]` which already causes the optimizer to loop indefinitely?